add bounds

NicolasColombi opened the pull request:
Changes proposed in this PR:

This PR add tropical cyclones basins bounds as a dictionary named `BASINS_BOUNDS` in 0-360 coordinates, following STORM dataset bounds definitions.

This addition can be used to determine whether a track originated in that basin simply by using shapely functionalities:

```
point = Point(lat, lon)
True_or_False = polygon.contains(point)
```

Particularly useful to decrease the memory overload during global wind fields computations by computing fields basin wide.

Preview of the bounds:
![image](https://github.com/user-attachments/assets/3b8a6671-2f31-4bee-a873-faa261d29003)


### PR Author Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] Correct target branch selected (if unsure, select `develop`)
- [ ] Descriptive pull request title added
- [ ] Source branch up-to-date with target branch
- [ ] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [ ] [Tests][testing] updated
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

chahank commented:
Simple yet useful addition!

I would suggest two things:
1- make it a data class instead
2- make it clear that these are the STORM basin definitions. These can vary substantially, and hence cannot be defined as a definitive default

Also, I am not sure to understand the point `Particularly useful to decrease the memory overload during global wind fields computations by computing fields basin-wide.`? 

In what way does this work beyond what is already possible?

chahank commented:
Thanks for the explanation! I am a bit confused though, because for each track only the centroids near the track are considered (and they are loaded one by one into memory). 

Could you maybe post an example of a code that shows this increase in memory efficiency? That would certainly be useful to see.

NicolasColombi commented:
I see, then I might have misinterpreted Sam's words. I haven't tested yet if this indeed increases efficiency (it was my hypothesis given the above), but I will try soon. Also, if you know already areas in the Winfield computation that could be optimized memory-wise, let me know. As my memory is currently exploding using global exposure at 450 arcsec with only 200 tracks (>50GB on euler) and >150GB at 150 arcsec resolution.

NicolasColombi commented:
@chahank I converted it to a dataclass (unsure whether this is how you intended it) and added a method `split_by_basin()` which creates 6 TCTtracks instances (1 per basin) from 1 TCTtrack instance.

chahank commented:
Thanks! Although the idea of the data class is precisely to not have a macro variable like BASIN. The idea is to have the data in that class directly. Also, the idea is that you can then clearly explain that this is an arbitrary boarder for basins based on one particular publication. 

I am not sure to understand the purpose of the method `split_by_basin()`. Can you please provide an example?

chahank commented:
Exactly. The data class is there to contain the data. A global variable like BASIN is hard to manage and document, and is usually quite obscure to the user as it does not appear in the documentation.

chahank commented:
@emanuel-schmid aaa no of course, sorry for the confusion! I meant an `Enum` class.

NicolasColombi commented:
@emanuel-schmid makes more sense now! Let me know if it is implemented as intended. For consistency, we may also consider adding the basin pressure to that class ?

Thanks @ValentinGebhart. I agree, I switched to -180,+180 as it is consistent with the tracks coordinate range. I had to split the "SP" basin into two polygons to avoid issues with the antimeridian (original reason why I chose longitude between 0-360).

Before I write the test for `subset_by_basin()`, let me know if you have comments on the function.

emanuel-schmid commented:
@NicolasColombi many thanks. I've got too questions about the subset_by_basin method

1) wouldn't it be preferable to return an array with the basins instead of a dictionary with lists?
2) supposedly there are much more performant ways to do this than to loop over the data - is speed irrelevant here?

NicolasColombi commented:
> @NicolasColombi many thanks. I've got two questions about the subset_by_basin method
> 
> 1. wouldn't it be preferable to return an array with the basins instead of a dictionary with lists?
> 2. supposedly there are much more performant ways to do this than to loop over the data - is speed irrelevant here?

@emanuel-schmid 
For 1. don't you lose the information on which basin the TCTtracks object belongs to with an array ? I just picked a format where you can access the data by calling the basin name. But mybe there is a better option :upside_down_face:

For 2. Speed is not so much of an issue here, but if you know a more efficient way I am absolutely keen to modify the function. Can you advise on how to optimize it?

NicolasColombi commented:
> > @NicolasColombi many thanks. I've got two questions about the subset_by_basin method
> > 
> > 1. wouldn't it be preferable to return an array with the basins instead of a dictionary with lists?
> > 2. supposedly there are much more performant ways to do this than to loop over the data - is speed irrelevant here?
> 
> @emanuel-schmid For 1. don't you lose the information on which basin the TCTtracks object belongs to with an array ? I just picked a format where you can access the data by calling the basin name. But mybe there is a better option 🙃
> 
> For 2. Speed is not so much of an issue here, but if you know a more efficient way I am absolutely keen to modify the function. Can you advise on how to optimize it?

@emanuel-schmid Any updates ?

chahank commented:
Would it make sense to also add the bounds for the impact basins as defined in https://www.nature.com/articles/s41467-022-33918-1 ?

NicolasColombi commented:
> Would it make sense to also add the bounds for the impact basins as defined in https://www.nature.com/articles/s41467-022-33918-1 ?

Maybe, what do you have in mind ? But at the moment I personally do not see a direct application. So, we can either add it now without a clear idea, or we add it later if needed. I would personally opt for the latter, unless you had something in mind.

chahank commented:
> Maybe, what do you have in mind ? But at the moment I personally do not see a direct application. So, we can either add it now without a clear idea, or we add it later if needed. I would personally opt for the latter, unless you had something in mind.

I am a bit confused now. I thought your goal was to be able to load tracks for specific basins. You will also use the basins defined in  https://www.nature.com/articles/s41467-022-33918-1 . Hence, why would you not implement these?

NicolasColombi commented:
Because I do not need them, nor I see a use case aside from defining centroids, which I already have. But of course feel free to add them; I am sure someone will find it useful.

chahank commented:
Still confused, but ok.

Then please just choose a name that makes obvious that this is only the definition used in STORM (is this really defined by STORM, or did they use the definition from somewhere else?). Currently, it looks like the Basin is a universally defined set of boundaries.

NicolasColombi commented:
> Still confused, but ok.
> 
> Then please just choose a name that makes obvious that this is only the definition used in STORM (is this really defined by STORM, or did they use the definition from somewhere else?). Currently, it looks like the Basin is a universally defined set of boundaries.

Not really, STORM defined boundaries based on IBTracks, but did not use exactly the same. I wrote it clearly at the beginning of the class description that these boundaries were defined according to STORM. I think it is up to the user to check this information, especially since it is easy to find and basin boundaries are very similar across different publications. Adding "STORM" to the class name would be confusing in my opinion.

emanuel-schmid commented on climada/hazard/tc_tracks.py:
```diff
@@ -322,6 +415,68 @@ def subset(self, filterdict):
+
+        Example:
+        --------
+        >>> tc = TCTracks.from_ibtracks("")
```
as an example this is suboptimal:
```ptyhon
>>> tc = TCTracks.from_ibtracks("")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'TCTracks' has no attribute 'from_ibtracks'
```

emanuel-schmid commented on climada/hazard/tc_tracks.py:
```diff
@@ -322,6 +415,68 @@ def subset(self, filterdict):
+        tracks_outside_basin: list = []
+        # Iterate over each tropical cyclone
+        for track in self.data:
+            lat, lon = track.lat.values[0], track.lon.values[0]
+            origin_point = Point(lon, lat)
+            point_in_basin = False
+
+            # Find the basin that contains the point
+            for basin in Basin:
+                if basin.value.contains(origin_point):
+                    basins_dict[basin.name].append(track)
+                    point_in_basin = True
+                    break
+
+            if not point_in_basin:
+                tracks_outside_basin.append(track.id_no)
```
@NicolasColombi as to 1. and 2.
I would first define a method that assigns basins to a track:
```python
BASINS_GDF = gpd.GeoDataFrame(
    {'basin': b, 'geometry': b.value} for b in Basin
)

def get_basins(track):  # this is the method I had in mind for 1. and I'd guess it could be a performance boost
    track_coordinates = GeoDataFrame(
        geometry=gpd.points_from_xy(track.lon, track.lat)
    )
    return track_coordinates.sjoin(BASINS_GDF , how='left', predicate='within').basin
```

emanuel-schmid replied:
Having that I'd write the track loop like this:
```python
for track in self.data:
    touched = get_basins(track).dropna().drop_duplicates()
    if touched.size:
        for basin in touched:
            basins_dict[basin].append(track)
    else:
        tracks_outside_basin.append(track)
```

chahank replied:
I have no strong opinion, but I would say the second makes more sense as a default. Ideally, just allow for both.

NicolasColombi replied:
Ok! I implemented Emanuel version and allowed the flexibility of choosing only the origin basin or not.

emanuel-schmid reviewed (commented):
updates 😄

chahank commented on climada/hazard/tc_tracks.py:
```diff
@@ -193,6 +195,102 @@
+    )
+
+
+BASINS_GDF = gpd.GeoDataFrame(
+    {"basin": b, "geometry": b.value} for b in Basin_bounds_storm
+)
```
This beats the purpose of the enum class a bit. Please do without it. This should be part of the method `get_basins`. The latter should also have a parameter `source` or similar, which at the moment can only have the value `STORM`. The method `subset_by_basin` should also have that.

chahank commented on climada/hazard/tc_tracks.py:
```diff
@@ -322,6 +420,59 @@ def subset(self, filterdict):
+
+        Returns
+        -------
+        dict_tc_basins : dict
+            A dictionary where the keys are basin names (e.g., "NA", "EP", "WP", etc.) and the
+            values are instances of the `TCTracks` class: effectively all tracks that intersected or originated
+            in each basin, depending on the argument "origin".
```
I think it makes much more sense to return for each basin a TCTrack object, and not a list of xarray datasets.

chahank commented on climada/hazard/tc_tracks.py:
```diff
@@ -322,6 +415,87 @@ def subset(self, filterdict):
+    def get_basins(track):
+        """Identify the tropical-cyclone basins crossed by a single track.
+
+        The function appends the latitude and longitude coordinates of the input track
+        into a GeoDataFrame of point geometries and performs a spatial join with the
+        global BASINS_GDF polygons (each representing a tropical-cyclone basin).
+        It returns the basin name(s) for every point along the track.
```
```suggestion
        Provides the basin name for every point along the track.
        The basins are defined according to the STORM definition:
        https://www.nature.com/articles/s41597-020-0381-2
        The names are:
        WP: West Pacific
        NA: North Atlantic
        NI: North Indian
        SP: South Pacific
        SI: South Indian
        EP: East Pacific
```
Please add the source of the basin definition.

chahank replied:
Correct?
